aidial_adapter_bedrock/llm/converse/adapter.py

```diff
@@ -0,0 +1,181 @@
+
+ConverseMessages = List[Tuple[ConverseMessage, Any]]
+
+
+class ConverseAdapter(ChatCompletionAdapter):
+    deployment: str
+    bedrock: Bedrock
+    storage: FileStorage | None
+
+    tokenize_text: Callable[[str], int] = default_tokenize_string
+    input_tokenizer_factory: Callable[
+        [ConverseDeployment, ConverseRequestWrapper],
+        Callable[[ConverseMessages], Awaitable[int]],
+    ]
+    support_tools: bool
+    partitioner: Callable[[ConverseMessages], List[int]] = (
+        turn_based_partitioner
+    )
+
+    async def _discard_messages(
+        self, params: ConverseRequestWrapper, max_prompt_tokens: int | None
+    ) -> Tuple[DiscardedMessages | None, ConverseRequestWrapper]:
+        if max_prompt_tokens is None:
+            return None, params
+
+        discarded_messages, messages = await truncate_prompt(
+            messages=params.messages.list,
+            tokenizer=self.input_tokenizer_factory(self.deployment, params),
+            keep_message=keep_last,
+            partitioner=self.partitioner,
+            model_limit=None,
+            user_limit=max_prompt_tokens,
+        )
+
+        return list(
+            params.messages.to_original_indices(discarded_messages)
+        ), ConverseRequestWrapper(
+            messages=ListProjection(
+                omit_by_indices(messages, discarded_messages)
+            ),
+            system=params.system,
+            inferenceConfig=params.inferenceConfig,
+            toolConfig=params.toolConfig,
+        )
+
+    async def count_prompt_tokens(
+        self, params: ModelParameters, messages: List[DialMessage]
+    ) -> int:
+        converse_params = await self.construct_converse_params(messages, params)
+        return await self.input_tokenizer_factory(
+            self.deployment, converse_params
+        )(converse_params.messages.list)
+
+    async def count_completion_tokens(self, string: str) -> int:
+        return self.tokenize_text(string)
+
+    async def compute_discarded_messages(
+        self, params: ModelParameters, messages: List[DialMessage]
+    ) -> DiscardedMessages | None:
+        converse_params = await self.construct_converse_params(messages, params)
+        discarded_messages, _ = await self._discard_messages(
+            converse_params, params.max_prompt_tokens
+        )
+        return discarded_messages
+
+    def get_tool_config(self, params: ModelParameters) -> ConverseTools | None:
+        if params.tool_config and not self.support_tools:
+            raise ValidationError("Tools are not supported")
+        return (
+            to_converse_tools(params.tool_config)
+            if params.tool_config
+            else None
+        )
+
+    async def construct_converse_params(
+        self,
+        messages: List[DialMessage],
+        params: ModelParameters,
+    ) -> ConverseRequestWrapper:
+        system_prompt_extraction = extract_converse_system_prompt(messages)
```

Comment:
Maybe a better name would be `prepare_converse_request`?

Reply:
I do not agree here,`prepare` is too generic naming.

When user see `extract` and `system_prompt`, it gives him clear understanding, what's happening, combined with `non_system_messages` in result dataclass